File: haskell/E2_1.hs
-- runhaskell E2_1.hs

-- Integer is a bigint and will not overflow
fib :: Integer -> Integer
fib n = fibR n 0 1
  where
    fibR 0 a _ = a
    fibR n a b = fibR (n - 1) b (a + b)

-- $ just wraps the rest of the line in ()
main :: IO ()
main = putStrLn $ show $ map fib [0..100]

File: scala/E2_1.scala
// scala E2_1.scala

object E2_1 {
  def fib(n: Int): Int = {
    @scala.annotation.tailrec
    def fibR(n: Int, a: Int, b: Int): Int =
      if (n <= 0) a else fibR(n - 1, b, a + b)
    fibR(n, 0, 1)
  }

  def main(args: Array[String]) = println(1 to 10 map fib)
}
